packages/in_app_purchase/ios/Classes/FIAPaymentQueueHandler.m

```diff
@@ -31,21 +32,28 @@ - (instancetype)initWithQueue:(nonnull SKPaymentQueue *)queue
                         updatedDownloads:(nullable UpdatedDownloads)updatedDownloads {
   self = [super init];
   if (self) {
-    self.queue = queue;
-    self.transactionsUpdated = transactionsUpdated;
-    self.transactionsRemoved = transactionsRemoved;
-    self.restoreTransactionFailed = restoreTransactionFailed;
-    self.paymentQueueRestoreCompletedTransactionsFinished = restoreCompletedTransactionsFinished;
-    self.shouldAddStorePayment = shouldAddStorePayment;
-    self.updatedDownloads = updatedDownloads;
-    self.transactionsSetter = [NSMutableDictionary new];
-    [queue addTransactionObserver:self];
+    _queue = queue;
+    _transactionsUpdated = transactionsUpdated;
+    _transactionsRemoved = transactionsRemoved;
+    _restoreTransactionFailed = restoreTransactionFailed;
+    _paymentQueueRestoreCompletedTransactionsFinished = restoreCompletedTransactionsFinished;
+    _shouldAddStorePayment = shouldAddStorePayment;
+    _updatedDownloads = updatedDownloads;
+    _transactionsSetter = [NSMutableDictionary dictionary];
   }
   return self;
 }
 
-- (void)addPayment:(SKPayment *)payment {
+- (void)startObservingPaymentQueue {
+  [_queue addTransactionObserver:self];
+}
+
+- (BOOL)addPayment:(SKPayment *)payment {
+  if (self.transactionsSetter[payment.productIdentifier]) {
```

Comment:
Can user buy multiple same "consumable"? This seems to prevent that to happen.

Reply:
I also don't see a test for this.
Another option is to do it on the dart side, which is easier testable.

Reply:
Normally, a good user experience would be app showing a blocking UI when a transaction is still pending such as Amazon. Also, if we allow the user to purchase the same product more than once, it might be very hard for apps that allows multi-login to deliver products to the right user.

Let's say app uses Google Sign-in as app's account system, user signed-in with account foo, purchased product A and closed the app before transaction finishes. At this time, transaction in the SKPaymentQueue doesn't have a transaction identifier yet, thus, there is no way for the app to establish an one-to-one mapping between the user account and the transaction. (applicationUserName is nil when state changes.) Next time, user launches the app and signed-in with account bar. The product will be delivered to bar instead of foo. If app has offers, then there will be more too many edge cases to handle on the app side. Limiting one pending transaction per product identifier will not hurt user experience a lot but dramatically decrease the complicity of the issue both us and client engineer need to solve.

I will add a test case on the dart side, thanks for calling this out, good catch!

Reply:
Good catch! Added logic and its associated test.